src/librustc/ty/subst.rs

```diff
@@ -541,6 +541,8 @@ impl<'a, 'gcx, 'tcx> SubstFolder<'a, 'gcx, 'tcx> {
         let opt_ty = self.substs.get(p.idx as usize).map(|k| k.unpack());
         let ty = match opt_ty {
             Some(UnpackedKind::Type(ty)) => ty,
+            Some(UnpackedKind::Const(ty::Const { ty, ..})) => ty,
+            None => source_ty,
```

Comment:
This is incredibly suspect, but it is the branch that does fail (I think after adding the one above).

Reply:
I think this hides the issue without actually fixing it. What we want to do is forbid const parameters to have type parameters as their types. These used to be forbidden, but it hasn't been merged yet.